modules/extensions/src/commands/index.ts

```diff
@@ -1,35 +1,151 @@
+   */
+  description?: string;
+
+  /**
+   * The command's icon. This is a relative path to the icon that will be displayed next to the label in the Commandbar.
+   *
+   * For example, if `public` is your statically served directory and you have an icon at `./public/icons/cmd.svg`, you would set this to `/icons/cmd.svg`
+   */
+  icon?: string;
 };
 
-export function Command({ data, commands, run }: CommandArgs) {
-  let wsCmd: CommandArgs = {
-    data,
-  };
+export type ActionCommandArgs = BaseCommandArgs & {
+  /**
+   * The function that will be called when someone choses to run this command.
+   */
+  run: Run;
+};
 
-  if (commands && run) {
+export type ContextCommandArgs = BaseCommandArgs & {
+  /**
+   * The function that will be called to compute 'subcommands' for this command. The subcommands can be any number of commands,
+   * and even computed dynamically in response to the different arguments passed to this function or some other external state.
+   */
+  commands: CommandsFn;
+};
+
+export type CommandArgs = ActionCommandArgs | ContextCommandArgs;
+
+/**
+ * This validates a command. We make sure that exactly one of `commands` or `run` is defined, and every other argument is serializable.
+ */
+function validateCommand(cmdArgs: unknown): asserts cmdArgs is CommandArgs {
+  // Make sure cmdArgs is object
+  if (typeof cmdArgs !== "object") {
+    throw new Error("Command arguments must be an object");
+  }
+  if (cmdArgs === null) {
+    throw new Error("Command arguments must not be null");
+  }
+
+  // Make sure it contains `commands` or `run`
+  if (!("commands" in cmdArgs) && !("run" in cmdArgs)) {
+    throw new Error("One of `commands` or `run` must be defined");
+  }
+
+  // But not both
+  if (
+    "commands" in cmdArgs &&
+    cmdArgs.commands &&
+    "run" in cmdArgs &&
+    cmdArgs.run
+  ) {
     throw new Error("Only one of `commands` or `run` must be defined");
   }
 
-  if (commands) {
-    wsCmd.data.type = "context";
-    wsCmd.commands = async (...args) => {
-      return proxy(await commands(...args));
-    };
-  } else if (run) {
-    wsCmd.data.type = "action";
-    wsCmd.run = run;
+  // And when provided, they must always be a function
+  if ("commands" in cmdArgs && typeof cmdArgs.commands !== "function") {
+    throw new Error("`commands` must be a function");
   }
 
-  return proxy(wsCmd);
+  if ("run" in cmdArgs && typeof cmdArgs.run !== "function") {
+    throw new Error("`run` must be a function");
+  }
+
+  // Make sure all other arguments are serializable
+  for (let entry of Object.entries(cmdArgs)) {
+    if (entry[0] === "commands" || entry[0] === "run") {
+      continue;
+    }
+
+    try {
+      JSON.stringify({ [entry[0]]: entry[1] });
+    } catch (e) {
+      throw new Error(`Command argument '${entry[0]}' is not serializable`);
+    }
+  }
+}
+
+export function Command(cmdArgs: CommandArgs) {
```

Comment:
We talked about maybe making this a simple interface, and expect POJOs to be passed around.

Reply:
Will follow up with this